delegation/delegated.gno

```diff
@@ -0,0 +1,252 @@
+	gDelegatedToken = &DelegatedToken{
+		staker:      staker,
+		token:       token,
+		delegatedTo: make(map[std.Address]std.Address),
+		balances:    make(map[std.Address]*u256.Uint),
+		totalSupply: u256.Zero(),
+		name:        name,
+		symbol:      symbol,
+	}
+}
+
+// GetStaker returns the addrress of the staker that this delegated token wrapper uses.
+func GetStaker() std.Address {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+	return gDelegatedToken.staker
+}
+
+// GetDelegatedTo returns the address to whom the owner is delegated.
+func GetDelegatedTo(owner std.Address) std.Address {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	delegatee, exists := gDelegatedToken.delegatedTo[owner]
+	if !exists {
+		// if not explicitly delegated, return owner
+		return owner
+	}
+
+	return delegatee
+}
+
+func Delegate(owner, to std.Address) {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	currDelegatee := GetDelegatedTo(owner)
+
+	// if the new delegation is different from the current one
+	if currDelegatee != to {
+		gDelegatedToken.delegatedTo[owner] = to
+
+		std.Emit(
+			"Delegation",
+			"from", owner.String(),
+			"to", to.String(),
+		)
+	}
+}
+
+// Deposit transfers the approved amount of the staked token to this contract
+// and mints an GRC20 representing the staked amount.
+func Deposit() {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	caller := std.GetOrigCaller()
+	token := gDelegatedToken.token
+
+	// get the approved amount
+	allowance := token.Allowance(caller, std.CurrentRealm().Addr())
+	if allowance == 0 {
+		panic("no allowance")
+	}
+
+	err := token.TransferFrom(caller, std.GetOrigPkgAddr(), allowance)
+	if err != nil {
+		panic("transfer failed: " + err.Error())
+	}
+
+	// mint equivalent amount of delegated tokens
+	amount := u256.NewUint(allowance)
+	mintDelegatedTokens(caller, amount)
+
+	std.Emit(
+		"Deposit",
+		"from", caller.String(),
+		"amount", amount.ToString(),
+	)
+}
+
+// DepositAmount deposites a specific amount of tokens
+func DepositAmount(amount *u256.Uint) {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	caller := std.GetOrigCaller()
+	token := gDelegatedToken.token
+
+	// check allowance
+	allowance := token.Allowance(caller, std.GetOrigPkgAddr())
+	if u256.NewUint(allowance).Cmp(amount) < 0 {
+		panic("insufficient allowance")
+	}
+
+	// transfer tokens from caller to this contract
+	err := token.TransferFrom(caller, std.GetOrigPkgAddr(), amount.Uint64())
+	if err != nil {
+		panic("transfer failed: " + err.Error())
+	}
+
+	// mint equivalent amount of delegated tokens
+	mintDelegatedTokens(caller, amount)
+
+	std.Emit(
+		"DepositAmount",
+		"from", caller.String(),
+		"amount", amount.ToString(),
+	)
+}
+
+// Withdraw withdraws the entire staked balance from the contract for the caller
```

Comment:
Can we make it to be able to withdraw partially?

Reply:
`WithdrawAmount` does that, but I'll modify it so that it can be handled by a single `Withdraw` function.

Reply:
https://github.com/gnoswap-labs/governance/pull/9/commits/681e41d9529d86d8ee8adacbe9d59a148474fbec